django/db/models/query_utils.py

```diff
@@ -150,6 +150,18 @@ def deconstruct(self):
             kwargs["_negated"] = True
         return path, args, kwargs
 
+    @property
+    def fields(self):
+        def determine_fields(filter_expr):
+            if isinstance(filter_expr, Q):
+                return filter_expr.fields
+            field = filter_expr[0].split(LOOKUP_SEP)[0]
+            return {field}
```

Comment:
It also needs to account about usage of boolean expressions such as lookups (e.g. `Q(Exact('field', 134))`); not all rhs are tuples of the form `(lookup: str, value: Any)` as `Q.__init__(*args)` also ends up in `Q.children`.

https://github.com/django/django/blob/a69b0e9cfe0af7cd2deaf55c069453c4c4598604/django/db/models/query_utils.py#L48-L50

Reply:
Ah yes I forgot about lookups 👍

Reply:
Taking a cursory look at lookups, something like `Exact('field', 134)`… we don't know whether `'field'` is a value or a reference to a field name. 🤔 I see where we'd start needing some magic.

Reply:
Oh actually, this would be `Exact(F('field'), 134)` sorry for the confusion.

Reply:
ha yes just as I walked afk I realised `Exact('field', 134)` wouldn't work with constraint check because there's no model to reference and came back to test that.

Nice so if constraints must use `F` that makes things easier.

Reply:
> Nice so if constraints must use F that makes things easier.

Right you can likely use a similar walk based on `hasattr(expr, "flatten")` with `isinstance(expr, F)` checks.

Reply:
@charettes 

I've pushed another version that makes use of `flatten()` to retrieve all `Q` and `F` instances and retrieve field references from both. It didn't occur to me how useful flatten would be until you'd pointed it out 🙂

It currently flattens twice as I've written it in a functional style to demonstrate collecting F & Q fields separately but the end result may need to switch to a single loop or comprehension - whichever is more readable while still respecting the generator.

I've also pushed an additional commit to move the field check to `validate_constraints()` as I think code should be set unconditionally. I still agree that entire condition along with `len(fields) == 1` check may be better within `UniqueConstraints.validate()` although this is the least impactful change but 🤷‍♂️ What are your thoughts on doing that here?